File: kotlin/src/test/kotlin/com/solana/mpp/client/SessionTest.kt
package com.solana.mpp.client

import com.solana.mpp.crypto.Base58
import com.solana.mpp.crypto.Ed25519
import com.solana.mpp.crypto.MemorySigner
import com.solana.mpp.crypto.PublicKey

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Client-side ActiveSession voucher lifecycle. Mirrors the Rust spine
 * (client/session.rs) and the Go session_test.go: monotonic cumulative,
 * Ed25519 verification against the authorized signer, and action builders.
 */
class SessionTest {
    private val channel = PublicKey(ByteArray(32) { 9 })
    private val signer = MemorySigner.fromSeed(ByteArray(32) { 3 })

    private fun session(): ActiveSession = ActiveSession.create(channel, signer)

    @Test
    fun authorizedSignerMatchesSignerPublicKey() {
        val s = session()
        assertEquals(Base58.encode(signer.publicKeyBytes), s.authorizedSigner)
        assertEquals(channel.toBase58(), s.channelIdString)
    }

    @Test
    fun signIncrementAdvancesWatermarkAndVerifies() {
        val s = session()
        val v1 = s.signIncrement(100)
        assertEquals(100L, s.cumulative)
        assertEquals("100", v1.data.cumulative)

        val v2 = s.signIncrement(50)
        assertEquals(150L, s.cumulative)
        assertEquals("150", v2.data.cumulative)

        // Ed25519 verifies against the authorized signer over the 48 byte message.
        val message = v2.data.messageBytes()
        assertEquals(48, message.size)
        val sig = Base58.decode(v2.signature)
        assertTrue(Ed25519.verify(signer.publicKeyBytes, message, sig))
    }

    @Test
    fun prepareDoesNotAdvanceWatermark() {
        val s = session()
        s.signIncrement(100)
        val prepared = s.prepareIncrement(25)
        assertEquals("125", prepared.data.cumulative)
        // Watermark unchanged until recorded.
        assertEquals(100L, s.cumulative)
        s.recordVoucher(prepared)
        assertEquals(125L, s.cumulative)
    }

    @Test
    fun voucherMustStrictlyIncrease() {
        val s = session()
        s.signVoucher(100)
        assertFailsWith<com.solana.mpp.protocol.MppException> { s.signVoucher(100) }
        assertFailsWith<com.solana.mpp.protocol.MppException> { s.signVoucher(50) }
    }

    @Test
    fun incrementMustBePositive() {
        val s = session()
        assertFailsWith<com.solana.mpp.protocol.MppException> { s.signIncrement(0) }
        assertFailsWith<com.solana.mpp.protocol.MppException> { s.prepareIncrement(-1) }
    }

    @Test
    fun voucherActionWrapsSignedIncrement() {
        val s = session()
        val action = s.voucherAction(200)
        assertTrue(action is SessionAction.Voucher)
        assertEquals("200", action.payload.voucher.data.cumulative)
        assertEquals(200L, s.cumulative)
    }

    @Test
    fun openActionPush() {
        val s = session()
        val action = s.openAction(1_000_000, "opensig")
        assertTrue(action is SessionAction.Open)
        val payload = action.payload
        assertEquals(SessionMode.PUSH, payload.mode)
        assertEquals(channel.toBase58(), payload.channelId)
        assertEquals("1000000", payload.deposit)
        assertEquals(s.authorizedSigner, payload.authorizedSigner)
        assertEquals("opensig", payload.signature)
    }

    @Test
    fun openPaymentChannelAction() {
        val s = session()
        val action = s.openPaymentChannelAction(2_000_000, "payer", "payee", "mint", 99, 45, "opensig")
        val payload = (action as SessionAction.Open).payload
        assertEquals(99L, payload.salt)
        assertEquals(45, payload.gracePeriod)
        assertEquals("payer", payload.payer)
    }

    @Test
    fun openPullAction() {
        val s = session()
        val action = s.openPullAction(5_000_000, "wallet", "approvesig")
        val payload = (action as SessionAction.Open).payload
        assertEquals(SessionMode.PULL, payload.mode)
        assertEquals(channel.toBase58(), payload.tokenAccount)
        assertEquals("wallet", payload.owner)
    }

    @Test
    fun topUpAction() {
        val s = session()
        val action = s.topUpAction(9_000_000, "topupsig")
        assertTrue(action is SessionAction.TopUp)
        assertEquals("9000000", action.payload.newDeposit)
    }

    @Test
    fun closeActionWithoutFinalVoucher() {
        val s = session()
        val action = s.closeAction(null)
        assertNull((action as SessionAction.Close).payload.voucher)
    }

    @Test
    fun closeActionWithFinalVoucher() {
        val s = session()
        s.signIncrement(100)
        val action = s.closeAction(50)
        val payload = (action as SessionAction.Close).payload
        assertNotNull(payload.voucher)
        assertEquals("150", payload.voucher.data.cumulative)
        assertEquals(150L, s.cumulative)
    }

    @Test
    fun explicitExpiryPropagatesToVouchers() {
        val s = ActiveSession.createWithExpiry(channel, signer, 1234)
        val v = s.signIncrement(10)
        assertEquals(1234L, v.data.expiresAt)
        s.setExpiresAt(5678)
        val v2 = s.signIncrement(10)
        assertEquals(5678L, v2.data.expiresAt)
    }
}

File: go/protocols/mpp/client/session_test.go
package client

import (
	"crypto/ed25519"
	"testing"

	solana "github.com/gagliardetto/solana-go"

	"github.com/solana-foundation/pay-kit/go/protocols/mpp/intents"
)

// seededSigner returns a deterministic signer from a fixed 32-byte seed,
// mirroring the make_signer helper in the Rust client/session tests.
func seededSigner(seed byte) solana.PrivateKey {
	b := make([]byte, 32)
	for i := range b {
		b[i] = seed
	}
	return solana.PrivateKey(ed25519.NewKeyFromSeed(b))
}

func newTestSession() *ActiveSession {
	channel := solana.NewWallet().PublicKey()
	return NewActiveSession(channel, seededSigner(42))
}

func TestSignIncrementIncreasesCumulative(t *testing.T) {
	s := newTestSession()
	if s.Cumulative() != 0 {
		t.Fatalf("initial cumulative = %d", s.Cumulative())
	}
	v, err := s.SignIncrement(100)
	if err != nil {
		t.Fatal(err)
	}
	if s.Cumulative() != 100 || v.Data.Cumulative != "100" || *v.Data.Nonce != 1 {
		t.Fatalf("increment mismatch: cum=%d v=%+v", s.Cumulative(), v.Data)
	}
}

func TestSignVoucherRejectsNonIncreasing(t *testing.T) {
	s := newTestSession()
	if _, err := s.SignIncrement(100); err != nil {
		t.Fatal(err)
	}
	if _, err := s.SignVoucher(100); err == nil {
		t.Fatalf("expected non-increasing rejection")
	}
	if _, err := s.SignVoucher(50); err == nil {
		t.Fatalf("expected non-increasing rejection")
	}
}

func TestPrepareAndRecordAreSeparate(t *testing.T) {
	s := newTestSession()
	prepared, err := s.PrepareIncrement(75)
	if err != nil {
		t.Fatal(err)
	}
	if s.Cumulative() != 0 {
		t.Fatalf("prepare should not advance watermark")
	}
	if err := s.RecordVoucher(prepared); err != nil {
		t.Fatal(err)
	}
	if s.Cumulative() != 75 {
		t.Fatalf("record should advance to 75, got %d", s.Cumulative())
	}
	if err := s.RecordVoucher(prepared); err == nil {
		t.Fatalf("re-record should fail")
	}
}

func TestNonceIncrementsPerVoucher(t *testing.T) {
	s := newTestSession()
	v1, _ := s.SignIncrement(10)
	v2, _ := s.SignIncrement(10)
	if *v1.Data.Nonce != 1 || *v2.Data.Nonce != 2 {
		t.Fatalf("nonce sequence mismatch: %d %d", *v1.Data.Nonce, *v2.Data.Nonce)
	}
}

// Golden vector: a voucher signed by the session must verify under the session
// signer's public key over the program Borsh voucher bytes. This locks the
// signing-byte layout used cross-language without a live chain.
func TestSignedVoucherVerifiesAgainstSignerKey(t *testing.T) {
	channel := solana.NewWallet().PublicKey()
	signer := seededSigner(7)
	s := NewActiveSession(channel, signer)
	voucher, err := s.SignIncrement(250)
	if err != nil {
		t.Fatal(err)
	}
	message, err := voucher.Data.MessageBytes()
	if err != nil {
		t.Fatal(err)
	}
	sig, err := solana.SignatureFromBase58(voucher.Signature)
	if err != nil {
		t.Fatal(err)
	}
	pub := signer.PublicKey()
	if !ed25519.Verify(ed25519.PublicKey(pub[:]), message, sig[:]) {
		t.Fatalf("voucher signature did not verify against signer key")
	}
	if voucher.Data.ChannelID != channel.String() {
		t.Fatalf("voucher channel id mismatch")
	}
}

func TestOpenActionFields(t *testing.T) {
	s := newTestSession()
	action := s.OpenAction(1_000_000, "txsig123")
	if action.Action != intents.ActionOpen || action.Open == nil {
		t.Fatalf("open action shape mismatch")
	}
	if action.Open.Mode != intents.SessionModePush || action.Open.Deposit != "1000000" || action.Open.Signature != "txsig123" {
		t.Fatalf("open action fields mismatch: %+v", action.Open)
	}
	if action.Open.AuthorizedSigner != s.AuthorizedSigner() {
		t.Fatalf("authorized signer mismatch")
	}
}

func TestOpenPullActionFields(t *testing.T) {
	s := newTestSession()
	action := s.OpenPullAction(5_000_000, "wallet123", "approvesig")
	if action.Open.Mode != intents.SessionModePull || action.Open.ApprovedAmount != "5000000" || action.Open.Owner != "wallet123" {
		t.Fatalf("pull open fields mismatch: %+v", action.Open)
	}
	if action.Open.TokenAccount != s.ChannelIDStr() {
		t.Fatalf("pull token account should be channel id")
	}
}

func TestTopUpActionFields(t *testing.T) {
	s := newTestSession()
	action := s.TopUpAction(5_000_000, "topuptx")
	if action.Action != intents.ActionTopUp || action.TopUp.NewDeposit != "5000000" || action.TopUp.Signature != "topuptx" {
		t.Fatalf("topup action mismatch: %+v", action.TopUp)
	}
}

func TestCloseActionWithAndWithoutVoucher(t *testing.T) {
	s := newTestSession()
	action, err := s.CloseAction(nil)
	if err != nil {
		t.Fatal(err)
	}
	if action.Close.Voucher != nil {
		t.Fatalf("close without increment should omit voucher")
	}

	s2 := newTestSession()
	if _, err := s2.SignIncrement(100); err != nil {
		t.Fatal(err)
	}
	final := uint64(50)
	action2, err := s2.CloseAction(&final)
	if err != nil {
		t.Fatal(err)
	}
	if action2.Close.Voucher == nil || action2.Close.Voucher.Data.Cumulative != "150" {
		t.Fatalf("close with final increment mismatch")
	}
}

func TestSetExpiresAt(t *testing.T) {
	channel := solana.NewWallet().PublicKey()
	s := NewActiveSessionWithExpiry(channel, seededSigner(42), 1234)
	v, _ := s.PrepareIncrement(10)
	if v.Data.ExpiresAt != 1234 {
		t.Fatalf("expiry = %d, want 1234", v.Data.ExpiresAt)
	}
	s.SetExpiresAt(5678)
	v2, _ := s.PrepareIncrement(10)
	if v2.Data.ExpiresAt != 5678 {
		t.Fatalf("expiry = %d, want 5678", v2.Data.ExpiresAt)
	}
}
